src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/FrameworkEventSource.cs

```diff
@@ -0,0 +1,199 @@
+                    descrs[4].Reserved = 0;
+                    descrs[5].DataPointer = (IntPtr)(&arg6);
+                    descrs[5].Size = 4;
+                    descrs[5].Reserved = 0;
+                    WriteEventCore(eventId, 6, descrs);
+                }
+            }
+        }
+
+        // optimized for common signatures (used by the ThreadTransferSend/Receive events)
+        [NonEvent]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Concurrency", "CA8001", Justification = "This does not need to be correct when racing with other threads")]
+        private unsafe void WriteEvent(int eventId, long arg1, int arg2, string arg3)
+        {
+            if (IsEnabled())
+            {
+                if (arg3 == null) arg3 = "";
+                fixed (char* string3Bytes = arg3)
+                {
+                    EventSource.EventData* descrs = stackalloc EventSource.EventData[3];
+                    descrs[0].DataPointer = (IntPtr)(&arg1);
+                    descrs[0].Size = 8;
+                    descrs[0].Reserved = 0;
+                    descrs[1].DataPointer = (IntPtr)(&arg2);
+                    descrs[1].Size = 4;
+                    descrs[1].Reserved = 0;
+                    descrs[2].DataPointer = (IntPtr)string3Bytes;
+                    descrs[2].Size = ((arg3.Length + 1) * 2);
+                    descrs[2].Reserved = 0;
+                    WriteEventCore(eventId, 3, descrs);
+                }
+            }
+        }
+
+        // optimized for common signatures (used by the EndGetRequestStream event)
+        [NonEvent]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Concurrency", "CA8001", Justification = "This does not need to be correct when racing with other threads")]
+        private unsafe void WriteEvent(int eventId, long arg1, bool arg2, bool arg3)
+        {
+            if (IsEnabled())
+            {
+                EventSource.EventData* descrs = stackalloc EventSource.EventData[3];
+                descrs[0].DataPointer = (IntPtr)(&arg1);
+                descrs[0].Size = 8;
+                descrs[0].Reserved = 0;
+                descrs[1].DataPointer = (IntPtr)(&arg2);
+                descrs[1].Size = 4;
+                descrs[1].Reserved = 0;
+                descrs[2].DataPointer = (IntPtr)(&arg3);
+                descrs[2].Size = 4;
+                descrs[2].Reserved = 0;
+                WriteEventCore(eventId, 3, descrs);
+            }
+        }
+
+        [Event(30, Level = EventLevel.Verbose, Keywords = Keywords.ThreadPool | Keywords.ThreadTransfer)]
+        public void ThreadPoolEnqueueWork(long workID)
+        {
+            WriteEvent(30, workID);
+        }
+
+        [NonEvent]
+        public unsafe void ThreadPoolEnqueueWorkObject(object workID)
+        {
+            // convert the Object Id to a long
+            ThreadPoolEnqueueWork((long)*((void**)Unsafe.AsPointer(ref workID)));
```

Comment:
I realize it's been like this, but this seems strange, given that the GC could move the object between the enqueue and dequeue, and their "ID"s thus wouldn't line up.

Reply:
I agree it's odd but that's how it's tracked now

Reply:
And the high 32 bits will be random garbage on 32-bit.

.NET Framework had a lot more FrameworkEventSource events, but we have dropped most of them in .NET Core. @vancem @kouvel Are these few threadpool-related events that are still left in the FrameworkEventSource useful?

Reply:
I'm not sure if anyone uses them, I haven't used them in the past. The thread pool one is a very verbose event fired for each work item enqueue and dequeue. They could be useful in some cases, though I don't have the historical context. It looks like the code would dereference a stack pointer into a `void*` and convert it to a `long`, wouldn't that be zero-extended on 32-bit?

Reply:
> wouldn't that be zero-extended on 32-bit?

Yes, you are right.

Reply:
These two events Enqueue and Dequeue, are being used (and are useful).   Yes, using an object reference as an ID is not stable as you point out, but at the time these were put in (they were put in as servicing changes), it was the best we could do (adding another piece of state to old the ID was too invasive).    It is also true that tools like PerfView CAN in fact track GCs and what updates were done so while the ID is not stable it CAN be stabilized by the tool.   

We should not remove it (although thinking about doing it better is certainly a possibility).

Reply:
Ok. I wonder how tools like PerfView deal with the fact that the same object can now be queued multiple times, and even live in the queues multiple times concurrently. Something to confirm works as expected.